src/coreclr/jit/importer.cpp

```diff
@@ -2831,6 +2831,86 @@ GenTree* Compiler::impImportLdvirtftn(GenTree*                thisPtr,
     return call;
 }
 
+//------------------------------------------------------------------------
+// impStoreNullableFields: create a Nullable<T> object and store
+//    'hasValue' (always true) and the given value for 'value' field
+//
+// Arguments:
+//    nullableCls - class handle for the Nullable<T> class
+//    value       - value to store in 'value' field
+//
+// Return Value:
+//    A local node representing the created Nullable<T> object
+//
+GenTree* Compiler::impStoreNullableFields(CORINFO_CLASS_HANDLE nullableCls, GenTree* value)
+{
+    // We still make some assumptions about the layout of Nullable<T> in JIT so let's assert them
+    static_assert_no_msg(OFFSETOF__CORINFO_NullableOfT__hasValue == 0);
+    unsigned hasValOffset = OFFSETOF__CORINFO_NullableOfT__hasValue;
+    unsigned valueOffset  = info.compCompHnd->getFieldOffset(info.compCompHnd->getFieldInClass(nullableCls, 1));
+
+    // Create an empty Nullable<T> object on stack
+    unsigned resultTmp = lvaGrabTemp(true DEBUGARG("Nullable<T> tmp"));
+    lvaSetStruct(resultTmp, nullableCls, false);
+    GenTreeLclVar* result       = impCreateLocalNode(resultTmp DEBUGARG(0));
+    GenTree*       hasValueAddr = gtNewLclAddrNode(resultTmp, hasValOffset);
+    GenTree*       valueAddr    = gtNewLclAddrNode(resultTmp, valueOffset);
+
+    // Now store 'hasValue' and 'value' fields
+    GenTree* hasValueStore = gtNewStoreIndNode(TYP_UBYTE, hasValueAddr, gtNewIconNode(1));
+    GenTree* valueStore;
+
+    if (varTypeIsStruct(value))
+    {
+        CORINFO_CLASS_HANDLE unboxCls = info.compCompHnd->getTypeForBox(nullableCls);
+        valueStore                    = gtNewStoreBlkNode(typGetObjLayout(unboxCls), valueAddr, value);
+    }
+    else
+    {
+        valueStore = gtNewStoreIndNode(value->TypeGet(), valueAddr, value);
+    }
```

Comment:
I don't see why we can't just use `gtNewStoreLclFldNode` here. It has an overload that takes a `ClassLayout*`.

Note that I think the non-struct case is wrong. Both cases need to get the underlying type from the EE since it may not necessarily match the JIT type on `value` (e.g. the value could be a small indirection that still needs a larger store).